Feat: import edn / json export of Logseq

cnrpman opened the pull request:
Can be accessed in the onboarding import UI
<img width="258" alt="image" src="https://user-images.githubusercontent.com/9862022/170504846-93ae13eb-9ea3-483f-8339-f93cf4c81d82.png">


Recover information from exported `edn` / `json` in the best effort.
- no FS required
- add options toward `handler/page/create!` and `handler/editor/insert-block-tree` to maintain uuid during reconstruction
- uuids are persisted on demand
- basic error capturing - skip pages with errors (e.g. missing block ref) and prompt

Known issue of the current export format:
- assets are dropped
- title properties are duplicated in dupilcated (won't fix since title properties are going to be deprecated)

vhscom commented:
> * duplicated properties in `:block/properties` and the children content

Is this referring to when the page block has properties how the first child formats those properties as content for the page. It would seem dropping all block properties is not necessary is this special case is handled, as individuals may be using their pages to define RDF triples and other metadata which is as important as the content itself.

cnrpman commented:
@vhscom Yes, so we need to make the exported `:block/properties` more reliable.
Parsing the properties from content is a temporary workaround.

tiensonqin commented:
@cnrpman @vhscom https://github.com/logseq/logseq/commit/7bf613348f0d7f90dedf1bc38875c70b3c90d4c8 removes all the properties from `:block/content`.

tiensonqin commented:
@cnrpman Let's resolve the conflicts and merge this PR!

logseq-cldwalker commented on src/main/frontend/handler/external.cljs:
```diff
@@ -95,3 +99,117 @@
+    pages))
+
+(defn- import-from-tree!
+  "Not rely on file system - backend compatible.
```
Is this so it can work on the web client easily?

cnrpman replied:
Yes, it's the design consideration.

cnrpman replied:
It should also works on server backend in future

logseq-cldwalker replied:
Nice

logseq-cldwalker commented on src/main/frontend/handler/editor.cljs:
```diff
@@ -1954,22 +1962,31 @@
                 (assoc fst-block :block/level (:block/level block)))))))
 
-(defn insert-block-tree-after-target
+(defn insert-block-tree
```
These operations are interesting. It looks like these fns and the underlying `frontend.modules.outliner.*` namespaces could be made nbb-logseq compatible if we want some day

logseq-cldwalker commented on src/main/frontend/handler/external.cljs:
```diff
@@ -95,3 +99,117 @@
+  "Not rely on file system - backend compatible.
+   tree-translator-fn: translate exported tree structure to the desired tree for import"
+  [data tree-translator-fn]
+  (when-let [_repo (state/get-current-repo)]
```
Should this be deleted if we aren't using the repo?

logseq-cldwalker reviewed (commented):
@cnrpman Nice work! Left some minor, optional comments. To QA this, I exported [this small test graph](https://github.com/cldwalker/logseq-query/tree/main/test-notes) and then imported it. I noticed two differences:
* Page test.content-search.md was renamed to test%2Fcontent-search.md
* The imported page has the property `title:: test/content-search` that the original page didn't have

I don't know if these are bugs but thought I should share.

It would be nice if we had a test that loaded a graph, exported it, imported it and then compared the before graph with the after graph. We [have the ability to load graphs into a test db](https://github.com/logseq/logseq/blob/1857f884d2012b41ed95736f3cec33e66b9219ad/src/test/frontend/handler/repo_test.cljs#L14-L18). Maybe this is better as another task?

Feel free to merge if the two differences above aren't bugs

cnrpman commented on src/main/frontend/components/external.cljs:
```diff
@@ -1,3 +1,5 @@
+;; deprecated by the onboarding import panel frontend.components.onboarding.setups
+
 (ns frontend.components.external
```
@xyhp915

logseq-cldwalker replied:
This doesn't have to hold up the PR but would be good to know why it's still around

tiensonqin commented:
@cnrpman @logseq-cldwalker 
> I noticed two differences:

> Page test.content-search.md was renamed to test%2Fcontent-search.md
The imported page has the property title:: test/content-search that the original page didn't have

I think it's a bug, a page with the title `test.content-search` should be saved to `test.content-search.md` and have the title property `title:: test.content-search` in the markdown file.

cnrpman commented:
> @cnrpman Nice work! Left some minor, optional comments. To QA this, I exported [this small test graph](https://github.com/cldwalker/logseq-query/tree/main/test-notes) and then imported it. I noticed two differences:
> 
>     * Page test.content-search.md was renamed to test%2Fcontent-search.md
> 
>     * The imported page has the property `title:: test/content-search` that the original page didn't have
> 
> 
> I don't know if these are bugs but thought I should share.
> 
> It would be nice if we had a test that loaded a graph, exported it, imported it and then compared the before graph with the after graph. We [have the ability to load graphs into a test db](https://github.com/logseq/logseq/blob/1857f884d2012b41ed95736f3cec33e66b9219ad/src/test/frontend/handler/repo_test.cljs#L14-L18). Maybe this is better as another task?
> 
> Feel free to merge if the two differences above aren't bugs

@tiensonqin @logseq-cldwalker I just play this. I clone the repo and load without any export / import, it's parsed as namespace page
<img width="552" alt="image" src="https://user-images.githubusercontent.com/9862022/173050200-0a619abe-697d-4c7f-af7a-dd260eb4cffb.png">
I think this is the default behavior for pages without `title` property included (for now, at least)
And we will switch into parsed as `test.content-search` when most users are using the version that using URL encode for file name.
( https://github.com/logseq/logseq/pull/5250#issuecomment-1120451059 )

cnrpman commented:
@logseq-cldwalker 
> This retrieves the db N times for N datoms. Would be more efficient to pass the db above

I'm not a pro on datascript 🤣  Hope I'm writing a neat query this time:
https://github.com/logseq/logseq/blob/26131e7bff8c319beb59eb51fa598052a9799b82/src/main/frontend/db/model.cljs#L1357-L1365
[DIFF](https://github.com/logseq/logseq/pull/5446/commits/26131e7bff8c319beb59eb51fa598052a9799b82#diff-4077c98d1e4aed129217962916bb93e042afd5c8144c0bd4e8935b2aa7e491bdL1361-R1365)

**Benchmarking on Lambda**
10x faster
```
VM3263:5 new: 43.18994140625 ms
VM3264:5 old: 643.13427734375 ms
VM3265:5 old: 652.881103515625 ms
VM3266:5 new: 49.030029296875 ms
VM3267:5 new: 64.070068359375 ms
VM3268:5 old: 663.403076171875 ms
```
Producing same results:
``` clojure
(util/profile "new" (count (get-all-referenced-blocks-uuid)))
1322
cljs꞉frontend.db.model꞉> 
(util/profile "new" (count (get-all-referenced-blocks-uuid-new)))
1322
cljs꞉frontend.db.model꞉> 
(util/profile "old" (count (get-all-referenced-blocks-uuid)))
1322
cljs꞉frontend.db.model꞉> 
(util/profile "old" (count (get-all-referenced-blocks-uuid)))
1322
cljs꞉frontend.db.model꞉> 
(util/profile "new" (count (get-all-referenced-blocks-uuid-new)))
1322
cljs꞉frontend.db.model꞉> 
(util/profile "new" (count (get-all-referenced-blocks-uuid-new)))
1322
cljs꞉frontend.db.model꞉> 
(util/profile "old" (count (get-all-referenced-blocks-uuid)))
1322
```

logseq-cldwalker commented:
> For my understanding, this is the default behavior for pages without `title` property included.
> And we will switch into parsed as `test.content-search` when most users are updated to the Logseq version that using URL encode for file name, then we can drop the ugly `title` property.

I didn't quite understand this. Are namespaced files going to be eventually url encoded? Agree that a roadmap or even just a a quire task or two would be helpful

logseq-cldwalker reviewed (approved):
@cnrpman Thanks for addressing feedback! 👍 🎉 I would encourage an end-to-end test if this is functionality that will eventually be critical to users. Happy to help you with it later if you decide to try it. Don't fully understand the namespace page behavior so I will defer to @tiensonqin for approval on that

cnrpman commented:
> @cnrpman Thanks for addressing feedback! 👍 🎉 I would encourage an end-to-end test if this is functionality that will eventually be critical to users. Happy to help you with it later if you decide to try it. Don't fully understand the namespace page behavior so I will defer to @tiensonqin for approval on that

I will try adding e2e test in the next cycle :)

andelf commented:
Cannot access this when you already have graphs.
Better add this to the top-right menu -> Import.

cnrpman commented:
@andelf There is an entry point in the import panel (it's reusing the onboarding UI)
<img width="737" alt="image" src="https://user-images.githubusercontent.com/9862022/173585832-d96b45bb-55bc-4899-bc8c-b55a07ac8610.png">
